### DISCUSSION TITLE
Performance and scalability

### DISCUSSION THREAD

**Discussion by erotavlasme** [owncast/owncast#508] [Q&A]
Hi,
the project is very interesting. According to the [FAQ](https://owncast.online/docs/faq/) a dedicated machine 10 $/months is able to handle three different bitrates. What is the maximum number of users? 
Do you have any study about the scalability of the project in terms of CPU/RAM/Bandwidth?
Thank you

**Comment by gabek** [comment]
Unfortunately it's difficult to give you a concrete answer, because much of it depends on your requirements and configuration.  Also, if you decide to use S3 storage to take the bandwidth burden from your server, then you can scale your video pretty much indefinitely.  Otherwise your bandwidth requirements will scale linearly with the number of viewers you have, and will very based on the bitrate set in your stream.

CPU is generally the largest bottleneck, but can be configured and tweaked for your system.

Sorry it's such a vague answer, but "it depends" is often the answer for things :)

**Comment by erotavlasme** [comment]
Thank you for your reply. I made similar considerations, so in order to scale to thousands of users, even with a single resolution 480p or 720p, it is required a very powerful system not one of 10 $/months.
In my opinion, it would be useful a study about scalability.
Best Regards

**Comment by jeyemwey** [comment]
From a cpu- and memory stand point, the video conversion tasks are much more expensive than the distribution to n  users. The only thing that scales linearly with more viewers is the bandwidth costs. That’s why we offer you to offload this task to a S3 hoster like Amazon or backblaze, if your own uplink cannot handle it.

**Comment by novadeviator** [comment]
I too am trying to understand more clearly how is the load (cpu/ram/bw) actually distributed. Do I understand correctly, that CPU/RAM of the main machine works on transcoding the original stream - let's say original is 1080p/3500kbps, and is transcoded to two lower bitrates 720p/2Mbps and 480p/1Mbps. if this is then distributed to another s3 hoster what are requirements for let's say 300 viewers that would scale this successfuly? does more viewers hit the CPU on transcoding machine in any way? if the main machine is at 50% cpu for all three stream points, does it matter if there are 50 or 300 viewers in case of additional s3 hoster?

(my use case is a VPS droplet on Digital Ocean for original stream point + another Digital Ocean Spaces bucket)

**Comment by aaronpk** [comment]
Once the video is transcoded it's just a matter of service the static files, there's no additional load on the transcoding server regardless of the number of viewers.

**Reply by novadeviator** [reply]
that sounds great! thank you! will be testing it in the coming week.

**Reply by novadeviator** [reply]
did some tests tonight. and I have some problems with encoding, but that is not that important now, I will work something out. But I'm also trying to setup external storage with DigitalOcean spaces. The problem is, there is no example conf for s3 spaces, in the docs there's only one for MinIO. is this the same also for DigitalOcean spaces? I have setup everything as instructed by my owncast is unable to sent fragments to the bucket or so it seems. I used something like:
```
s3:
  enabled: true
  endpoint: XXXX.fra1.digitaloceanspaces.com
  #endpoint: https://fra1.digitaloceanspaces.com
  #servingEndpoint: https://colinspace.fra1.digitaloceanspaces.com/stream
  accessKey: XXXXXXXXXXXXXXXXXXXXXXX
  secret: XXXXXXXXXXXXXXXXXXXXXX
  bucket: stream 
  region: anywhere
```

please advise. :-/

**Reply by gabek** [reply]
There is a section for DigitalOcean: https://owncast.online/docs/s3/#digital-ocean-spaces

But in general most S3 providers will be similar.  `servingEndpoint` is not required.

**Reply by novadeviator** [reply]
yes, I have been on that section for a while... so I reread it again, and ... managed to come up with this WORKING configuration:
```
s3:
  enabled: true
  endpoint: fra1.digitaloceanspaces.com
  accessKey: <ACCESS KEY>
  secret: <SECRET>
  bucket: <NAME OF THE SPACE> 
  region: fra1
```
Owncast now populates the bucket with hls folder and ts files, BUT, it seems to be unable to get them back to the player. I have setup CORS like this:

![image](https://user-images.githubusercontent.com/744626/105917728-0d706380-6033-11eb-92d8-d4e2bb6e20cf.png)

My server is at colin.tmp.si:XXXXX - so, on a custom port. Does that make a difference? I cannot seem to trace the error. The player is just stuck on the first frame and it doesn't move. 

Do I need to add an allowed header in that section above for CORS?

BTW thank you so far for the answers, and quite amazing software...

**Reply by gabek** [reply]
I do think the CORS origin requires the port.  If possible try changing it to a wildcard if it's supported just to test it.

What are the actual errors you're seeing in your browser console?  It's likely CORS related, but hopefully it's throwing out exactly what the issue is.  I know @geekgonecrazy uses Digital Ocean, so maybe he can chime in with anything specific.

**Reply by novadeviator** [reply]
I changed it to wildcard (only *), that was the only other option - can't add a port or something. And the issues is still there.

Ok, browser console says 404 not found:

```
GET http://colin.tmp.si:18989/thumbnail.jpg?cb=1611704753
[HTTP/1.1 200 OK 533ms]

XHR GET http://colin.tmp.si:18989/api/status
[HTTP/1.1 200 OK 150ms]

XHR GET http://colin.tmp.si:18989/hls/fra1.digitaloceanspaces.com/colinspace/hls/1/stream.m3u8?cachebust=1611704678?cachebust=1611704684?cachebust=1611704689?cachebust=1611704694?cachebust=1611704699?cachebust=1611704704?cachebust=1611704710?cachebust=1611704716?cachebust=1611704722?cachebust=1611704727?cachebust=1611704732?cachebust=1611704737?cachebust=1611704742?cachebust=1611704747?cachebust=1611704752?cachebust=1611704757
[HTTP/1.1 404 Not Found 24ms]

VIDEOJS: WARN: Problem encountered with playlist 1-fra1.digitaloceanspaces.com/colinspace/hls/1/stream.m3u8. HLS playlist request error at URL: fra1.digitaloceanspaces.com/colinspace/hls/1/stream.m3u8. Switching to playlist 0-fra1.digitaloceanspaces.com/colinspace/hls/0/stream.m3u8. core-c3086d77.js:2601:47
XHRGEThttp://colin.tmp.si:18989/hls/fra1.digitaloceanspaces.com/colinspace/hls/0/stream.m3u8?cachebust=1611704678?cachebust=1611704684?cachebust=1611704694?cachebust=1611704704?cachebust=1611704716?cachebust=1611704727?cachebust=1611704737?cachebust=1611704747?cachebust=1611704757
[HTTP/1.1 404 Not Found 25ms]

VIDEOJS: WARN: Removing other playlists from the exclusion list because the last rendition is about to be excluded. core-c3086d77.js:2601:47
VIDEOJS: WARN: Problem encountered with playlist 0-fra1.digitaloceanspaces.com/colinspace/hls/0/stream.m3u8. HLS playlist request error at URL: fra1.digitaloceanspaces.com/colinspace/hls/0/stream.m3u8. Switching to playlist 1-fra1.digitaloceanspaces.com/colinspace/hls/1/stream.m3u8. core-c3086d77.js:2601:47
XHR GET http://colin.tmp.si:18989/api/status
```

**Reply by gabek** [reply]
Change your endpoint from `fra1.digitaloceanspaces.com` to `https://fra1.digitaloceanspaces.com`.

**Reply by novadeviator** [reply]
it works with * as origin.
that is probably not a good /safe/ practice?

L
